sqlalchemy-stubs/ext/orderinglist.pyi

```diff
@@ -1,25 +1,40 @@
-from typing import Any
+from collections.abc import Sequence
+from typing import Callable
+from typing import List
 from typing import Optional
+from typing import TypeVar
+from typing import Union
 
-def ordering_list(attr: Any, count_from: Optional[Any] = ..., **kw: Any): ...
+_T = TypeVar("T")
+OrderingFunc = Callable[[int, Sequence[_T]], int]
 
-class OrderingList(list):
-    ordering_attr: Any = ...
-    ordering_func: Any = ...
-    reorder_on_append: Any = ...
+def ordering_list(
+    attr: str,
+    /,
+    count_from: Optional[int] = ...,
+    ordering_func: Optional[OrderingFunc] = ...,
+    reorder_on_append: bool = ...,
+) -> Callable[[], OrderingList]: ...
+
+class OrderingList(List[_T]):
+    ordering_attr: str = ...
+    ordering_func: OrderingFunc = ...
+    reorder_on_append: bool = ...
     def __init__(
         self,
-        ordering_attr: Optional[Any] = ...,
-        ordering_func: Optional[Any] = ...,
+        ordering_attr: Optional[str] = ...,
+        ordering_func: Optional[OrderingFunc] = ...,
         reorder_on_append: bool = ...,
     ) -> None: ...
     def reorder(self) -> None: ...
-    def append(self, entity: Any) -> None: ...
-    def insert(self, index: Any, entity: Any) -> None: ...
-    def remove(self, entity: Any) -> None: ...
+    def append(self, entity: _T) -> None: ...
+    def insert(self, index: int, entity: _T) -> None: ...
+    def remove(self, entity: _T) -> None: ...
     def pop(self, index: int = ...): ...
-    def __setitem__(self, index: Any, entity: Any) -> None: ...
-    def __delitem__(self, index: Any) -> None: ...
-    def __setslice__(self, start: Any, end: Any, values: Any) -> None: ...
-    def __delslice__(self, start: Any, end: Any) -> None: ...
+    def __setitem__(self, index: Union[int, slice], entity: _T) -> None: ...
```

Comment:
This should have the same signature as list:

```python
    @overload
    def __setitem__(self, i: int, o: _T) -> None: ...
    @overload
    def __setitem__(self, s: slice, o: Iterable[_T]) -> None: ...
```

Reply:
I think it would be good for us to add our own `_SupportsIndex` in `util/compat.pyi` and use that here instead of using a version check